choose server_name at installation

gnouts opened the pull request:
## Problem
https://github.com/YunoHost-Apps/synapse_ynh/issues/92

## Solution
- Add a server_name variable during install
Installation seems ok, but I haven't tested federation nor upgrade/backup/restore.

## PR Status
- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Validation
---
*Minor decision*
- **Upgrade previous version** : 
- [x] **Code review** : Josué
- [x] **Approval (LGTM)** : Josué
- [x] **Approval (LGTM)** : JimboJoe
- **CI succeeded** : 
[![Build Status](https://ci-apps-hq.yunohost.org/jenkins/job/synapse_ynh%20PR119/badge/icon)](https://ci-apps-hq.yunohost.org/jenkins/job/synapse_ynh%20PR119/) 

When the PR is marked as ready to merge, you have to wait for 3 days before really merging it.

Josue-T commented on manifest.json:
```diff
@@ -29,6 +29,15 @@
+				"name": "server_name",
+				"type": "string",
+				"ask": {
+					"en": "Choose a name for Matrix accounts",
```
Maybe can you specify that this is the domain name for the matrix accounts.

Thatoo replied:
Have a look at what I did here, it's more precise but maybe a too long explanation. Something in between would be good. https://github.com/YunoHost-Apps/synapse_ynh/compare/master...Thatoo:patch-1

I also had change the couple question/answer about public server that I personally found currently unclear.

gnouts replied:
@Thatoo I'm fine with what you wrote

Josue-T replied:
I purpose to put something like that:
```suggestion
					"en": "Choose a domain name for your matrix user-ids. Note that the user-ids look like something like that : @user:domain.tld instead",
```

Thatoo replied:
Well, by starting with the word "choose", it sounds like we are completely free to choose whatever we want which is not the case. We have to remain in the main domain name zone of yunohost...
If yunohost is on _mydomain.tld_ and synapse is on _synapse.mydonain.tld_, the synapse server_name can be
_mydomain.tld_
or
_matrix.mydomain.tld_
but it can't be
_matrix.tld_ or _securechat.me_
...

gnouts replied:
So waht do you think of this one ?
`If your synapse domain is a subdomain, you can select your main domain name to have your matrix user-ids looking like @user:domain.org instead of @user:synapse.domain.org. Otherwise enter the same subdomain as above.`

Josue-T commented on scripts/install:
```diff
@@ -228,6 +229,7 @@ cp ../conf/log.yaml /etc/matrix-$app/log.yaml
 
 ynh_replace_string __APP__ $app "$homeserver_config_path"
 ynh_replace_string __DOMAIN__ $domain "$homeserver_config_path"
+ynh_replace_string __SERVER_NAME__ $server_name "$homeserver_config_path"
```
Note that you need to  upgrade also the Upgrade and config file as explained in line 222.

gnouts replied:
Sure. Is changing the version in the manifest enough to simulate an upgrade ?
In short : how to quickly test an upgrade ?

Josue-T replied:
Just do an app upgrade.

You can update the manifest but for now it's not mandatory

Josue-T commented on scripts/install:
```diff
@@ -32,6 +32,7 @@ report_stats="False"
 #=================================================
 
 domain=$YNH_APP_ARG_DOMAIN
+server_name=$YNH_APP_ARG_SERVER_NAME
```
I think you need to save this settings and use it in the upgrade and config scripts.

Thatoo commented:
A change in README.md  is also necessary. 
I did one here https://github.com/YunoHost-Apps/synapse_ynh/compare/master...Thatoo:patch-2#diff-04c6e90faac2675aa89e2176d2eec7d8

gnouts commented:
I did some quick changes. I had tested install, backup, restore and upgrade.
Install is still fine. Though my federation test wasn't succesful. I think my setup is the faulty one, not synapse.
Backup is fine, as it wasn't changed.
Upgrade failed, just like restore. 

Here are my logs : 
- for upgrade (with tried to restore before deleting synapse) : https://paste.yunohost.org/raw/qafiqilexi
- for restore : https://paste.yunohost.org/raw/piyitalaje

My synapse was on `matrix.test-ynh.local` and server_name was `test-ynh.local`

Josue-T commented:
>  Though my federation test wasn't succesful. 

Did you apply these settings https://github.com/YunoHost-Apps/synapse_ynh#access-by-federation

gnouts commented:
I'm testing on my local network, I didn't setup a DNS. I just modified /etc/hosts on my VM and on my production Yunohost, hoping they see each other. I guess that the issue.
I'll try later with a "real" server on a domain name.

Josue-T commented:
> I'm testing on my local network, I didn't setup a DNS. I just modified /etc/hosts on my VM and on my production Yunohost, hoping they see each other. I guess that the issue.
I'll try later with a "real" server on a domain name.

Yes probably to try theses things you need a real install.

Josue-T commented:
I also saw that you need to update the  [check_process](https://github.com/YunoHost-Apps/synapse_ynh/blob/master/check_process).
Your probaly need to add update to something like that:
```
;; General
    ; Manifest
        domain="domain.tld"    (DOMAIN)
        path="/_matrix/static/"	(PATH)
        is_public=1 (PUBLIC|public=1|private=0)
        server_name="domain2.tld"    (DOMAIN)
; Checks
```

Thatoo commented:
Good job @gnouts !
It's on the way.
Thank you.

Thatoo commented:
I was reading about Matrix/Synapse 1.0 and I kind of understood (maybe I misunderstood) that the way it will works with server_name will be different. There is a question of .well-known to write instead of changing DNS... I even had the feeling that with this version, it would not be necessary to have a dedicated (sub)name domaine.
Does anyone understand similarly than me or could correct me?
If so, it was maybe to soon to work on that and we better wait for stable specification of v1.0...

Josue-T commented:
> There is a question of .well-known to write instead of changing DNS

So this too, but I didn't saw exactly how does it work. I did found a documentation about the setup of this solution. But I should say I searched theses information a long time ago...

Thatoo commented:
Ok, I report here my discussion on the synaspe.matrix.org room.

> When I re-install, should I go for SRV or .well-known?
> I ask also the question because we worked on DNS solution for the yunohost packge here https://github.com/YunoHost-Apps/synapse_ynh/issues/92 but as it's not yet merged, we could work on the other solution

> so in many cases it's easier to use .well-known, that's why it's being recommended
> if you know what you are doing, go ahead, you can get valid TLS certs and so on

> The choice of SRV or .well-known is all about how you set up your DNS, your server and your certificates. SRV is easier for some cases, .well-known is easier for other cases. Once you get your system set up, it doesn't make any practical difference which you are using: the "advantages and disadvantages" are all in the setting-up stage, not making any difference to how well the system works once it is running.
> (Except if you care about federating with old, pre-0.99 servers. For them, you need a SRV. But I recommend you should not care about them.)
> And be sure to understand: you don't need a SRV or a .well-known at all if your server is accessible on your server_name domain on the matrix default ports. You only need them if you want to redirect to a different hostname or different ports.

And here is the documentation about this subject : https://github.com/matrix-org/synapse/blob/master/docs/federate.md

Thatoo commented on scripts/install:
```diff
@@ -329,7 +332,7 @@ chown $synapse_user:root -R /var/lib/matrix-$app
+chmod 600 /etc/matrix-$app/$server_name.signing.key
 setfacl -R -m user:turnserver:rX  /etc/matrix-$app
 setfacl -R -m user:turnserver:rwX  /var/log/matrix-$app
 
```
If we go for the .well-known option, we need to add some scripts line to add a json file (server_name.json) into /var/www/.well-known/$domain/server_name.json 
and add a redirection of https://$domain/.well-known/matrix/server to that server_name.json file into /etc/nginx/conf.d/$domain.d/name_server.conf
something like

```
if [[ $server_name != $domain ]]; then
    cat > /var/www/.well-known/${domain}/server_name.json <<EOF
   {
          "m.server": "$domain"
   }
EOF

  cat > /etc/nginx/conf.d/${domain}.d/name_server.conf <<EOF
  {
    location /.well-known/matrix/server {
        alias /var/www/.well-known/${domain}/server_name.json;
        }
  }
EOF
fi
```
I'm a beginner so if I misunderstood something and my script is incorrect, please don't hesitate to comment for me to learn and for the package to improve.

The interest of going through the .well-known system is that we can then write in Readme that the change in DNS is completely optional and if they don't feel comfortable with changing DNS, they better not do it.

Josue-T replied:
I'll see to implement the .well-known support. It's also useful for the custom port case.

But it's important to make this change on an other PR, if you want to start on that.

Josue-T replied:
Anyway thanks for your useful informations.

Thatoo replied:
You're welcome.
I was thinking that maybe @gnouts would like to implement directly the .well-known system instead of the DNS system and as it's the PR for server_name, I thought it was ok to do it here.
It's only a question of some script lines to add (in install and upgrade scripts) and a modification of Readme to write that DNS is not mandatory anymore...
If an other PR is necessary then  I guess we better wait that @gnouts finish this one and it gets merged for good.

Thatoo replied:
I have to say that I'm waiting for this feature with impatience to be able to reinstall my synapse server on my family yunohost server.
Everything is like firstname@lastname.xz except matrix which is @firstname:synapse.lastname.xz and I find that not user friendly at all.
E2E cross-signing and this feature are the two missing features for me right now.

Thatoo replied:
Hello all,

I have been working on the .well-known solution and, with the help of synapse dev room on matrix, I managed to make it works.
Those who wants to have a look at my work, it's here : https://github.com/Thatoo/synapse_ynh/blob/Thatoo-choose-server-name/scripts/install
Just after install, without changing anything on DNS, the federation can make the relation between my server_name address and my synapse domain going to https://$server_name/.well-known/matrix/server address. However, it is not yet working properly because I guess the 8448 port is only open for the app synapse for entry coming for $domain adn it is closed for entry coming for $server_name address. Is that correct?

I guess I have to change something about that

```
# Open this port
yunohost firewall allow TCP $synapse_tls_port > /dev/null 2>&1
yunohost firewall allow Both $turnserver_tls_port > /dev/null 2>&1
yunohost firewall allow Both $turnserver_alt_tls_port > /dev/null 2>&1

# Store opened ports
ynh_app_setting_set $app synapse_port $port
ynh_app_setting_set $app synapse_tls_port $synapse_tls_port
ynh_app_setting_set $app turnserver_tls_port $turnserver_tls_port
ynh_app_setting_set $app turnserver_alt_tls_port $turnserver_alt_tls_port
ynh_app_setting_set $app cli_port $cli_port
```

Maybe add something to tell to open for this app on both address, $domain and $server_name.

Thatoo replied:
Should I make a PR already for us to work on that?

gnouts replied:
@Thatoo I'm closing this discussion as I think you've made your own branch by now and GitHub block merging if there's unresolved discussion :)

Thatoo commented:
I did it. It's working!
I'm currently talking with tulir from the synapse room with my @test:test.leprette.fr id on my matrix.test.leprette.fr synapse test server.
Install with server_name and .well-known works and federate!
I'll send you screenshot later and I'll PR.
If you want, I'm letting this server open tonight so you can ping me there tonight to test. Then I'll go back to my stable server (put back redirection of port...).
As soon as the v1.0 come I'll try the upgrade script too to see if it works also with server_name and .well-known.

Thatoo commented:
> I did some quick changes. I had tested install, backup, restore and upgrade.
> Install is still fine. Though my federation test wasn't succesful. I think my setup is the faulty one, not synapse.
> Backup is fine, as it wasn't changed.
> Upgrade failed, just like restore.
> 
> Here are my logs :
> 
>     * for upgrade (with tried to restore before deleting synapse) : https://paste.yunohost.org/raw/qafiqilexi
> 
>     * for restore : https://paste.yunohost.org/raw/piyitalaje
> 
> 
> My synapse was on `matrix.test-ynh.local` and server_name was `test-ynh.local`

@gnouts , as I explain in here, https://github.com/YunoHost-Apps/synapse_ynh/pull/136,  it failed because you where missing this following line in script/upgrade
` ynh_replace_string __SERVER_NAME__ $server_name "$homeserver_config_path"`
You added this line in other scripts but you forgot it in this one.
Now it is working. By the way, your work is federating.
I only added the support of .well-known so that admins don't need to add any SRV record in their DNS, it works out of the box.

gnouts commented:
@Thatoo  Thanks for your work. I thought I would have much time to finish this earlier...
w/e I added what you suggested and my tests are good so far.
Installation is ok
Upgrade is ok
Backup is ok
Restore is ok

I used a real server this time. I was able to federate !
Basically everything worked fine.

@Josue-T What do I need to do for this to be merge ?
I rebased my branch on testing. I guess we just need to close de discussion on the manifest.json.

Josue-T commented:
> @Josue-T What do I need to do for this to be merge ?

- You probably need to update the checkprocess file (https://github.com/YunoHost-Apps/synapse_ynh/blob/testing/check_process)
- The CI test need to pass (fixing the checkprocess should fix the actual issue
- After I will review the code. After somebody else from the apps teams  need to review the code before to merge this.

Josue-T commented on scripts/upgrade:
```diff
@@ -31,6 +31,7 @@ upstream_version=$(ynh_app_upstream_version)
 #=================================================
 
 domain=$(ynh_app_setting_get $app special_domain)
+server_name=$(ynh_app_setting_get $app server_name)
```
You need to support the retro compatibility. So in the case of the `server_name` is not defined, you need to define it.

You probably need to add a code like that:

```bash
if [ -z $server_name ]; then
   server_name=$domain
   ynh_app_setting_set $app server_name  $domain
fi

Josue-T commented on manifest.json:
```diff
@@ -29,6 +29,15 @@
+				"name": "server_name",
+				"type": "string",
+				"ask": {
+					"en": "If your synapse domain is a subdomain, you can choose a name for your Synapse server to have your matrix user-ids looking like @user:domain.org instead of @user:synapse.domain.org",
```
Maybe we could add the possibility to have the same server_name than the domain by default. I think that is a specific use case, so I purpose to do add a default value with some thing like "Same than the domain". By this the user which don't really know what is this can just leave the default value.

gnouts replied:
How do you propose to do that ?
We need to fix a default value, but that depends on each install. I don't know how to get the main domain, from inside the json :/

Josue-T replied:
I purpose one solution (but it's like hackish... because I don't think that we have actually the possibility to have a dynamic default value).

You add a default key as same as in line 50, and the value could be something like "Same than the domain".

In the install script you can add a code like that:

```bash
if [ $server_name == "Same than the domain" ]; then
    server_name=$domain
fi

gnouts replied:
I tried what you suggested. It worked for me.

Josue-T commented on scripts/upgrade:
```diff
@@ -31,6 +31,11 @@ upstream_version=$(ynh_app_upstream_version)
+if [ -z $server_name ]; then     # for retro compatibility
+   server_name=$domain
+   ynh_app_setting_set $app server_name  $domain
+fi
```
Sorry, can you put this in line 63

gnouts replied:
I tried upgrading from the actual master branch and it worked well

Thatoo commented:
Sure, you can close this discussion for me.
I'm waiting for one of our two PR to be merged into testing.

gnouts commented:
Everything looks good to me. I tested installation, upgrade, backup, restore, upgrade from current version.

Still, CI failed on "Migrate from self signed certificate to cert managed by Yunohost". The check_process is updated.

It was before my changes and Jenkins didn't trigger since. Dunno what to do ^^'

Thatoo commented:
Can you check the list in your first post? and make a call for review?

- [x] Code finished.
- [x] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

Josue-T commented:
> Everything looks good to me. I tested installation, upgrade, backup, restore, upgrade from current version. It's federating.
> 
> Still, CI failed on "Migrate from self signed certificate to cert managed by Yunohost". The check_process is updated.
> 
> It was before my last changes and Jenkins didn't trigger since. Dunno what to do ^^

Well, the issue come frome somewhere else. To solve this you just need to merge the last testing version into your branch and it should fix this issue.

Thatoo commented:
@gnouts 
Now that you have updated your branch from testing, is "Migrate from self signed certificate to cert managed by Yunohost" working?

Thatoo commented:
@gnouts any news?

gnouts commented:
> @gnouts any news?

well... I did not had any issues with certificate migration personnally. Still don't know what to do more :/

Thatoo commented:
@kay0u @Josue-T @JimboJoe 
Could you review it?
Can it be merged into testing?

gnouts commented on scripts/restore:
```diff
@@ -106,6 +106,7 @@ cp ../settings/conf/turnserver.conf "$coturn_config_path"
 ynh_replace_string __APP__ $app "$coturn_config_path"
 ynh_replace_string __TURNPWD__ $turnserver_pwd "$coturn_config_path"
 ynh_replace_string __DOMAIN__ $domain "$coturn_config_path"
+ ynh_replace_string __SERVER_NAME__ $server_name "$homeserver_config_path"
```
@Thatoo sorry I think you're viewing an outdated fie, I don't have it on local nor on remote source

Thatoo replied:
Well, I remember I wrote that sentence many month ago but not yesterday. Maybe I didn't validated it at that time and yesterday I did without paying attention. My mistake. Sorry.

Josue-T commented on README.md:
````diff
@@ -41,11 +41,11 @@ If your server name is identical to the domain on which synapse is installed, an
 
 ```
-_matrix._tcp.<yourdomain.com> <ttl> IN SRV 10 0 <port> <server.name>
+_matrix._tcp.<server_name-you-choose-at-install> <ttl> IN SRV 10 0 <port> <domain-or-subdomain-of-synapse>
````
```suggestion
_matrix._tcp.<server_name.tld> <ttl> IN SRV 10 0 <port> <domain-or-subdomain-of-synapse.tld>
```

JimboJoe commented:
Can be merged in 3 days.
